p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/FtpTransportProtocolHandler.java

```diff
@@ -0,0 +1,198 @@
+        final FTPClient client = getClient(uri);
+
+        if (!client.hasFeature(FTPCmd.MDTM)) {
+            return -1;
+        }
+
+        final FTPFile file = client.mdtmFile(uri.getPath());
+
+        if (file == null) {
+            throw new IOException("Could not retrieve the last modification timestamp for " + uri);
+        }
+
+        return file.getTimestampInstant().toEpochMilli();
+    }
+
+    @Override
+    public File getFile(final URI uri) throws IOException {
+        final File localFile = getLocalFile(uri);
+
+        if (cacheConfig.isOffline()) {
+            if (!localFile.isFile()) {
+                throw new IOException("Maven is offline and the requested URL " + uri + " does not exist locally");
+            }
+
+            return localFile;
+        }
+
+        final FTPClient client = getClient(uri);
+        final String remotePath = uri.getPath();
+        final FTPFile remoteFile = getRemoteFile(client, remotePath);
+        final boolean isRemoteMissing = remoteFile == null;
+
+        if (localFile.isFile() && (isRemoteMissing || !mustRefresh(localFile, remoteFile))) {
+            return localFile;
+        }
+
+        if (isRemoteMissing) {
+            logger.warn("Could not find file " + remotePath);
+            return null;
+        }
+
+        final File parent = FileUtils.createParentDirectories(localFile);
+        final File tempFile = File.createTempFile("download", ".tmp", parent);
+
+        try (final InputStream is = client.retrieveFileStream(remotePath)) {
+            try (final FileOutputStream os = new FileOutputStream(tempFile)) {
+                IOUtils.copyLarge(is, os);
+            }
+        } catch (final IOException e) {
+            tempFile.delete();
+            throw e;
+        }
+
+        if (!client.completePendingCommand()) {
+            throw new IOException(String.format("Error retrieving file %s. %s", remotePath, client.getReplyString()));
+        }
+
+        if (localFile.isFile()) {
+            FileUtils.forceDelete(localFile);
+        }
+
+        FileUtils.moveFile(tempFile, localFile);
+        localFile.setLastModified(remoteFile.getTimestampInstant().toEpochMilli());
+        return localFile;
+    }
+
+    @PreDestroy
+    public void dispose() {
+        for (final Map.Entry<String, FTPClient> entry : clients.entrySet()) {
+            try {
+                entry.getValue().logout();
+                entry.getValue().disconnect();
+            } catch (final FTPConnectionClosedException e) {
+                // Connection already closed by the host
+            } catch (final IOException e) {
+                logger.error("Error disconnecting from host " + entry.getKey(), e);
+            }
+        }
+    }
+
+    protected FTPClient createFtpClient() {
+        final FTPClient client = new FTPClient();
+        client.setControlKeepAliveTimeout(Duration.ofSeconds(15));
+        client.setAutodetectUTF8(true);
```

Comment:
Usually it is good to use passive mode (as some firewalls blocking FTP traffic otherwise). 
Probably one can use a system property to control this (with defaulting to passive mode).

Reply:
@laeubi yup, I now set explicitly:
```
client.enterLocalPassiveMode();
client.setFileType(FTP.BINARY_FILE_TYPE);
```

Reply:
This seems a good way.